### PR TITLE
fix(perf): improve performance

### PR DESCRIPTION
**Pull request by megheaiulian**
Improve searching and looping performance.

### CODE REVIEW

File: cosmoz-default-tree.js [hunk 1]
```diff
@@ -103,14 +106,17 @@ export class DefaultTree extends Tree {
...
+	 */
 	_searchNodes(propertyValue, options, nodes = this._roots) {
-		let results = [];
+		const results = [];
 
-		nodes.some(node => {
-			results = results.concat(this.search(node, propertyValue, options));
-			return options.firstHitOnly && results.length > 0;
-		});
+		for (const node of nodes) {
```
Comment (megheaiulian) [inline cosmoz-default-tree.js:113]: We mitigate some performance & possible memory heap increases by not creating (and discarding) an extra array at each loop step.

File: cosmoz-tree.js [hunk 2]
```diff
@@ -78,11 +85,13 @@ export class Tree {
...
 	 * @param {node} node (The node of which the property value should be returned of.)
 	 * @param {String} propertyName (The name of the property. e.g. "name")
-	*/
-	getProperty(node, propertyName) { // eslint-disable-line no-unused-vars
+	 */
+	getProperty(node, propertyName) {
 		throw new Error('Must be implemented in derived object');
 	}
 
+	/* eslint-enable no-unused-vars */
```
Comment (JockeCK) [inline cosmoz-tree.js:93]: forgotten comment?

Comment (megheaiulian) [reply cosmoz-tree.js:93]: No. I first disable and then re-enable the rule.

File: cosmoz-default-tree.js (hunk 1 above)
Comment (JockeCK) [reply cosmoz-default-tree.js:113]: if firstHitOnly is true, will not results array be empty? since we return res before it can push the res?

Comment (megheaiulian) [reply cosmoz-default-tree.js:113]: If firstHitOnly is true and if `res` (temporary results) are not empty than we return them. 
Same as before.

Comment (JockeCK) [reply cosmoz-default-tree.js:113]: Ah right, first hit, sorry not thinking.

**Pull request merged**